src/lib.rs

```diff
@@ -303,6 +303,9 @@ pub fn create(value: &Vec<u8>, data: &Vec<u8>) -> CreateResult {
 
 #[warn(non_snake_case)]
 pub fn calldata_copy(from: usize, length: usize) -> Vec<u8> {
+    // Ensure no overreading happens
+    assert!((calldata_size() - from) >= length);
```

Comment:
We could also consider changing these to return a rust Result object, which would be an error if it tries to read out of bounds.

That way issues can be caught and handled by the caller.